### ISSUE TITLE
Running the inform and estimate stages in different processes

### ISSUE THREAD

**Issue by joezuntz** [LSSTDESC/rail_attic#45]
Hi all,

Because the training data used by `inform` method will typically be of a relatively small size and run on a single node, whereas the estimation can (optionally) be done in parallel by many nodes, it would be useful to be able to split these two phases.

This could be done just by ensuring that future Estimator subclasses, like the ones that are currently there, can be pickled.

As an addendum - is the intention that the testfile specified in the estimator is used within the library to test algorithm performance?  Or is it intended to be used for the eventual full catalogs?  I think it's the former, but if not then it would also be useful to remove the requirement that this be specified in the init method.

Does this sound okay?  I can add a picklability test if so.

**Comment by joezuntz** [comment]
For example (copying this from an email for clarity), I'd like to do something like:

```
# in first (training) process
cls = Estimator._find_subclass("FZBoost")
pz = cls(base_config, run_dict)
pz.inform()
with open('model.pickle', 'wb') as f:
    pickle.dump(pz, f)
# finish process
```

Then in another process, later:

```
with open('model.pickle', 'rb') as f:
    pz = pickle.load(f)
 
for data in iter_chunk_hdf5_data(cat_file, chunk_size, name):
    pz_dict = pz.estimate(data)
    # save pz_dict data somewhere
```

**Comment by sschmidt23** [comment]
I think that there are two separate use cases, the two that you mention in particular, though @aimalz can say what he had in mind for sure.  It is currently coded for something like a data challenge, where we are testing all algorithms with a single training file and a single test file.  It is relatively straightforward to modify the set up to read in some form of pickled result of the training for each code for the second case where we want to generate large catalogs, as you say we just read in the trained model rather than running the inform stage.  The form will be different for different codes, so we may want to think a bit about whether any solution is generic enough to cover all use cases (it won't always be just a pickle file, in some cases it will be a set of parameters), though I believe it may just end up "hardcoded" in the estimate function for each subclass as to what exactly is read in.  I think we started to discuss this at some point, and one reason that I didn't implement it was because we were still discussing provenance tracking, and how to keep track of the setup used to create such a pickle file.

**Comment by johannct** [comment]
Ok it does sound like something we definitely need to think over and likely pursue, but it needs some thoughts and trial and error stage regarding implementation, based on a sufficiently complex use case. @joezuntz Do you have such a use case at hand? I think within TXPIPE you also gave thought to provenance..... anything that can be pulled over and used more globally?

**Comment by joezuntz** [comment]
I've put together an example of what I'd like to do in TXPipe, currently in PR with the relevant piece here:

https://github.com/LSSTDESC/TXPipe/pull/154/files?file-filters%5B%5D=.py#diff-c0cf418e3987d3276e501d3e8edd824795d7827a8736bb8bb672176750fd7b3c

In this particular case I track the provenance automatically using the general and automatic system I have in TXPipe, by including a provenance dictionary in the pickle file just before the pickled model. The full code is tied somewhat to TXPipe, but you could certainly adapt pieces of it if useful.

The code that does this is a bit scattered over TXPipe, but you can see it in these three places:
https://github.com/LSSTDESC/TXPipe/blob/master/txpipe/utils/provenance.py
https://github.com/LSSTDESC/TXPipe/blob/master/txpipe/data_types/base.py#L47
https://github.com/LSSTDESC/TXPipe/blob/master/txpipe/base_stage.py#L37

If useful I could easily split off the TXPipe-specific pieces into a library other code could use.  Let me know if that would be useful.

**Comment by joezuntz** [comment]
@johannct has pointed out that in my TXPipe code I'm using `Estimator._find_subclass`, and that if I'm allowed to do that then it should presumably be a public method and thus with no underscore - does that seem reasonable to change here?

**Comment by johannct** [comment]
either that or we propose a `__call__` interface : 
```
import os, sys

class A(object):
    _subclasses = {"os":os,"sys":sys}
    @classmethod
    def __call__(cls,name):
        return cls._find_class_with_name(name)

    @classmethod
    def _find_class_with_name(cls,name):
        return cls._subclasses[name]

a=A()('os')
b=A()('sys')
```
not sure it is worth it.... comment welcome

**Comment by joezuntz** [comment]
Can you have `__call__` as a class method?  Wouldn't that conflict with how `__init__` is called?

**Comment by johannct** [comment]
Ok so I looked over the txpipe code. The question now is : how do we proceed with datafile types and provenance, essentially what is in https://github.com/LSSTDESC/TXPipe/blob/master/txpipe/data_types/base.py
We probably need some sort of RFC higher up about creating a desc_data_types repos for pipelines to use, or a desc_base repo with data_types as a submodule or subpackage. What do people think here, before I discuss it with Heather, and the launch the RFC?

**Comment by johannct** [comment]
@joezuntz The first pair of parameters is the init part. Example with an init function : 
```
import os, sys

class A(object):
    _subclasses = {"os":os,"sys":sys}
    @classmethod
    def __call__(cls,name):
        return cls._find_class_with_name(name)

    @classmethod
    def _find_class_with_name(cls,name):
        return cls._subclasses[name]
    def __init__(self, status):
        pass

a=A('ok')('os')
b=A('ok')('sys')
```

so in the case of Estimator I would need to check that all goes well if I call `Estimator(base_config,run_config)(name)` That would be nicely compact, but maybe not worth it. Anyway, at best it is a treat, and even the original `_` question can be left aside for now. The RFC is a more important question.

**Comment by joezuntz** [comment]
Thanks for the review on the TXPipe @johannct ! 

Since the pickle file works for the current solution I'll use that for now in TXPipe, but will keep track of this repo and watch as any of these change, and as you evolve I can keep TX up to date.

- methods for serializing an Estimator class - pickle for now, evolve as needed
- name/mechanism of `_find_subclass`
- changing to QP output (will use `build_tables` and handle parallel write myself)

**Comment by sschmidt23** [comment]
We merged in a version with the pickled model as the default (though specific codes could write a custom load_pretrained_model function in their subclass that would use something other than a pickled model, e.g. a parameterized form).  
I'm dealing with switching to qp output in a separate issue (#33) that will be merged in sometime in the near future.  Since the inform option is in place and merged, I'm closing this issue.
